### ISSUE TITLE
Add constant value patch type in Composition

### ISSUE THREAD

**Issue by rozcietrzewiacz** [crossplane/crossplane#2013]
### Use Case

While building a resource catalog, we are leveraging Helm to template Compositions and CompositeResourceDefinitions for individual resources. On the input of such chart, we define upstream (provider) resource kind/apiVersion, some resource defaults (those go to `spec.forProvider` dierctly), some configurables (those go to `patches`). On top of that, we specify values that are global to the cluster, such as `accoundID` and `region`, that are required as part of some individual parameters for `base` resoures. The cluster wide parameters should be fixed in a Composition instead of being set at the creation of each CompositeResource.

We stumbled upon an issue with setting a cluster-wide value `region` in a unified manner. Most of the cloud resources have that setting under `spec.forProvider.region`, but not all - so we wanted to allow ourselves to define an override for the target resource spec path, where the Composition would inject the hadcoded `region` value. One example of such exception is AWS S3 bucket, that expects region restriction defined through ` spec.forProvider.locationConstraint`.

To allow an override, we tried something like this in composition template:
```yaml
{{- range .Values.supportedResources }}
---
apiVersion: apiextensions.crossplane.io/v1beta1                              
kind: Composition                                                            
metadata:                                                                    
  name: x{{ include "xxx.resourcePlural" . }}.xxx.com
spec:
  compositeTypeRef:                                      
    apiVersion: xxx.com/v1alpha1               
    kind: X{{- .kind }}                                  
  resources:                                             
    - base:                                              
        apiVersion: {{ .apiVersion }}                    
        kind: {{ .kind }}                                
        spec:                                            
          providerConfigRef:                             
            name: {{ $.Values.defaultProvider | quote }} 
          forProvider:                                   
            {{- if (empty .overrideRegionPath) }}        
            region: {{ $.Values.aws.region | quote }}
            {{- end }}
      patches:
        {{- with .overrideRegionPath }}                                                              
        - fromFieldPath: "kind" # Whatever, just put it there!                                       
          toFieldPath: {{ . | quote }}                                                               
          transforms:                                                                                
            - type: string                                                                           
              string:                                                                                
                fmt: {{ $.Values.aws.region | quote }}                                               
        {{- end }}
#... 
{{- end }}
```

This does not work however, because the `string.fmt` syntax does not include any `%s`, which results in an error from `Sprintf`.
Neither leaving out `fromFieldPath` nor setting it to an empty string worked either. 

### How could Crossplane help solve your problem?

Provide a means for Composition to feed a raw value to a specified JSONPath of the target base.

**Comment by mcavoyk** [comment]
I've come across a similar use-case, when patching labels from a XRD to each composed resource but also wanting some static labels. My workaround so far has been to use the map transform, but a value patch would be a better solution.

```yaml
        - fromFieldPath: "metadata.labels"
          toFieldPath: "metadata.labels"
        - fromFieldPath: "kind"
          toFieldPath: "metadata.labels[someLabel]"
          transforms:
            - type: map
              map:
                "CompositeResource": "someValue"
```

**Comment by rozcietrzewiacz** [comment]
Nice hack, @mcavoyk ! I totally forgot about `map` transform.

**Comment by negz** [comment]
I don't think I'm quite grokking these use cases. Can you help me understand why you can't include these static values as part of the base resource, rather than as a patch? e.g.

```yaml
resources:
- base:
    apiVersion: example.org/v1
    kind: Example
    spec:
      region: a-constant-region
```

```yaml
resources:
- base:
    apiVersion: example.org/v1
    kind: Example
    metadata:
      labels:
        someLabel: someValue
```

**Comment by negz** [comment]
Ah - I see now in @mcavoyk's example the issue is that you'd like to patch over the entire labels array, which would overwrite any static labels you had included in the base. We might be able to address that specific use case if we merged rather than overwrite objects when one was patched onto another.

**Comment by rozcietrzewiacz** [comment]
@negz To answer your question:
> why you can't include these static values as part of the base resource, rather than as a patch?

Because we are operating on a variable jsonPath, which can be different for different individual resource types we are including in our resource catalog (notice the `{{- range .Values.supportedResources }}` at the top of my code snippet)

**Comment by negz** [comment]
I just took a quick pass at implementing this, thinking it would be pretty simple and low risk to add the functionality. One quirk though is how we'd represent the source value - is it a string, an int, etc?

```yaml
patches:
- type: FromConstantValue
   fromConstantValue: example
   toFieldPath: spec.example
```

If we went with an API something like the above I don't know that there's a clean way to represent in the CRD schema that `fromConstantValue` could be anything. I'm fairly sure CRDs won't allow us to use an `interface{}` field but we could possibly make it a `runtime.RawExtension` - i.e. an opaque blob of YAML as far as the `Composition` schema was concerned. This would be interesting since it would make it possible for the value to be an object or an array as well as a simple type like an int or string, but would have pointy edges because we'd have no way to validate it against a schema.

We could also just force the constant value to be a `string` and use the new type conversion transform in https://github.com/crossplane/crossplane/pull/1994 if we wanted it to be some other value, but that feels like a bit of a hack.

**Comment by negz** [comment]
> we could possibly make it a `runtime.RawExtension`

Unfortunately this doesn't work - the generated CRD schema expects the field to be of type `object`.

**Comment by muvaf** [comment]
> If we went with an API something like the above I don't know that there's a clean way to represent in the CRD schema that fromConstantValue could be anything

This could be an option:
```yaml
patches:
- type: FromConstantValue
  constantValue:
    int: 64
#   string: "someval"
  toFieldPath: spec.example
```

> We could also just force the constant value to be a string and use the new type conversion transform

From the use cases above, I feel like there won't really be variance in the type even if the fieldpath varies. So, it may not be so useful to force everyone to use the additional transform. On the other hand, most of the time it's `string`. I'm not really sure, I can be convinced either way.

**Comment by rozcietrzewiacz** [comment]
Could it be possible to consume the `from` as string always and add some logic in runtime to dynamically infer the target data type from the target resource schema?

**Comment by negz** [comment]
@muvaf The API you propose is close to what I was thinking too, though as usual I'd like a type field to distinguish user intent - i.e.:

```yaml
patches:
- type: FromConstantValue
  constantValue:
    type: String
    string: foo
```

I also suspect that strings will be the most common type, but I'm definitely not advocating for the approach in which a number would be written as a string, requiring a transform function. Definitely just thinking out loud with that idea.

> Could it be possible to consume the from as string always and add some logic in runtime to dynamically infer the target data type from the target resource schema?

@rozcietrzewiacz We don't currently have any logic to make decisions based on the target resource schema, though that is something we'd like to investigate in future. Regardless, I definitely prefer being explicit rather than implicit in the Composition if we can.

**Comment by muvaf** [comment]
@negz proposed API looks good to me. `constantValue.type` indicates the type in Go whereas `patches[*].type` is something Crossplane-specific but I don't really think we can find a better name for that field other than `type`.

**Comment by muvaf** [comment]
We just got together with folks from the community and went through the use cases this patch type would be useful and it seems like many of them would be solved by https://github.com/crossplane/crossplane/issues/2099 on managed resource level or with support for template string with no variables in https://github.com/crossplane/crossplane/pull/2352. The label problem @mcavoyk was having is more related to how fieldpath library works, which does a replace operation rather then merge, so it could be solved by [this](https://github.com/crossplane/crossplane/issues/2211).

But there can be more use cases that we couldn't think of. So we decided to wait for a week for feedback from the community before proceeding in either way but hold the PRs for now.
